#include "Reports.h"
#include "ui_Reports.h"
#include "../core/DatabaseManager.h"
#include "../core/LogManager.h"
#include <QMessageBox>
#include <QFile>
#include <QTextStream>
#include <QDateTime>
#include <QDesktopServices>
#include <QUrl>
#include <QInputDialog>
#include <QDir>

Reports::Reports(QWidget *parent)
    : QWidget(parent)
    , ui(new Ui::Reports)
    , m_currentUser("admin")
{
    ui->setupUi(this);

    connect(ui->backButton, &QPushButton::clicked, this, &Reports::backRequested);
    connect(ui->generateButton, &QPushButton::clicked, this, &Reports::onGenerateReport);
    connect(ui->viewButton, &QPushButton::clicked, this, &Reports::onViewReport);
    connect(ui->deleteButton, &QPushButton::clicked, this, &Reports::onDeleteReport);
    
    refreshReports();
}

Reports::~Reports()
{
    delete ui;
}

void Reports::refreshReports()
{
    ui->reportListWidget->clear();
    
    QList<QVariantMap> reports = DatabaseManager::instance().listReports();
    
    for (const QVariantMap &report : reports) {
        int id = report["id"].toInt();
        QString createdAt = report["created_at"].toString();
        QString user = report["user"].toString();
        QString title = report["title"].toString();
        QString format = report["format"].toString();
        
        QString displayText = QString("[%1] %2 - %3 (%4)")
            .arg(createdAt)
            .arg(user)
            .arg(title)
            .arg(format);
        
        QListWidgetItem *item = new QListWidgetItem(displayText);
        item->setData(Qt::UserRole, id);
        item->setData(Qt::UserRole + 1, report["filepath"].toString());
        ui->reportListWidget->addItem(item);
    }
}

void Reports::onGenerateReport()
{
    // Get report title from user
    bool ok;
    QString title = QInputDialog::getText(this, "Generate Report",
                                         "Report Title:", QLineEdit::Normal,
                                         "Sample Report", &ok);
    if (!ok || title.isEmpty()) {
        return;
    }
    
    // Generate a sample report file
    QString reportsDir = LogManager::instance().getReportsDirectory();
    QString timestamp = QDateTime::currentDateTime().toString("yyyyMMdd_HHmmss");
    QString filename = QString("report_%1.txt").arg(timestamp);
    QString filepath = reportsDir + QDir::separator() + filename;
    
    QFile file(filepath);
    if (file.open(QIODevice::WriteOnly | QIODevice::Text)) {
        QTextStream out(&file);
        out << "Report: " << title << "\n";
        out << "Generated: " << QDateTime::currentDateTime().toString(Qt::ISODate) << "\n";
        out << "User: " << m_currentUser << "\n";
        out << "\n";
        out << "Sample report data:\n";
        out << "- Total users: " << DatabaseManager::instance().listUsers().size() << "\n";
        out << "- Log entries: " << DatabaseManager::instance().getRecentLogs(10).size() << " (recent)\n";
        out << "\n";
        out << "This is a sample report generated by the SandDrive system.\n";
        file.close();
        
        // Add to database
        QString err;
        if (DatabaseManager::instance().addReport(title, m_currentUser, filepath, "txt", &err)) {
            QMessageBox::information(this, "Generate Report", "Report generated successfully!");
            refreshReports();
        } else {
            QMessageBox::warning(this, "Generate Report", "Failed to save report metadata: " + err);
        }
    } else {
        QMessageBox::warning(this, "Generate Report", "Failed to create report file: " + filepath);
    }
}

void Reports::onViewReport()
{
    QListWidgetItem *selectedItem = ui->reportListWidget->currentItem();
    if (!selectedItem) {
        QMessageBox::information(this, "View Report", "Please select a report first");
        return;
    }
    
    QString filepath = selectedItem->data(Qt::UserRole + 1).toString();
    if (filepath.isEmpty() || !QFile::exists(filepath)) {
        QMessageBox::warning(this, "View Report", "Report file not found: " + filepath);
        return;
    }
    
    // Emit signal to show report viewer
    emit viewReportRequested(filepath);
}

void Reports::onDeleteReport()
{
    QListWidgetItem *selectedItem = ui->reportListWidget->currentItem();
    if (!selectedItem) {
        QMessageBox::information(this, "Delete Report", "Please select a report first");
        return;
    }
    
    int reportId = selectedItem->data(Qt::UserRole).toInt();
    QString filepath = selectedItem->data(Qt::UserRole + 1).toString();
    
    QMessageBox::StandardButton reply = QMessageBox::question(this, "Delete Report",
        "Are you sure you want to delete this report?",
        QMessageBox::Yes | QMessageBox::No);
    
    if (reply == QMessageBox::Yes) {
        // Delete from database
        QString err;
        if (DatabaseManager::instance().deleteReport(reportId, &err)) {
            // Delete file
            QFile::remove(filepath);
            QMessageBox::information(this, "Delete Report", "Report deleted successfully");
            refreshReports();
        } else {
            QMessageBox::warning(this, "Delete Report", "Failed to delete report: " + err);
        }
    }
}
